pushsync/pusher.go

```diff
@@ -0,0 +1,343 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pushsync
+
+import (
+	"context"
+	"encoding/hex"
+	"time"
+
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/spancontext"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/opentracing/opentracing-go"
+	olog "github.com/opentracing/opentracing-go/log"
+)
+
+// DB interface implemented by localstore
+type DB interface {
+	// subscribe to chunk to be push synced - iterates from earliest to newest
+	SubscribePush(context.Context) (<-chan storage.Chunk, func())
+	// called to set a chunk as synced - and allow it to be garbage collected
+	// TODO this should take ... last argument to delete many in one batch
+	Set(context.Context, chunk.ModeSet, ...storage.Address) error
+}
+
+// Pusher takes care of the push syncing
+type Pusher struct {
+	store    DB                     // localstore DB
+	tags     *chunk.Tags            // tags to update counts
+	quit     chan struct{}          // channel to signal quitting on all loops
+	pushed   map[string]*pushedItem // cache of items push-synced
+	receipts chan []byte            // channel to receive receipts
+	ps       PubSub                 // PubSub interface to send chunks and receive receipts
+}
+
+var (
+	retryInterval = 100 * time.Millisecond // seconds to wait before retry sync
+)
+
+// pushedItem captures the info needed for the pusher about a chunk during the
+// push-sync--receipt roundtrip
+type pushedItem struct {
+	tag         *chunk.Tag       // tag for the chunk
+	shortcut    bool             // if the chunk receipt was sent by self
+	firstSentAt time.Time        // first sent at time
+	lastSentAt  time.Time        // most recently sent at time
+	synced      bool             // set when chunk got synced
+	span        opentracing.Span // roundtrip span
+}
+
+// NewPusher contructs a Pusher and starts up the push sync protocol
+// takes
+// - a DB interface to subscribe to push sync index to allow iterating over recently stored chunks
+// - a pubsub interface to send chunks and receive statements of custody
+// - tags that hold the tags
+func NewPusher(store DB, ps PubSub, tags *chunk.Tags) *Pusher {
+	p := &Pusher{
+		store:    store,
+		tags:     tags,
+		quit:     make(chan struct{}),
+		pushed:   make(map[string]*pushedItem),
+		receipts: make(chan []byte),
+		ps:       ps,
+	}
+	go p.sync()
```

Comment:
i would remove this.
Since you are using this type from `swarm.go` it is expected to have a `Start` and `Stop` methods IMHO. Please rename `Close` to `Stop` and add a `Start` method that launches the pusher in a different goroutine and call that method from `swarm.go` after creating the instance. I would really like to see more standardization in the way we create and manage lifecycle of services in swarm

Reply:
this is the only component in `swarm` that launches a goroutine in the type constructor AFAIK. you are forcing a dependency on all of the plumbing around this thing to be set up correctly when you can just return the type and call the `go p.Sync()` in a separate statement.